elasticbeanstalk_versions Boto3 refactor

sidewinder12s opened the pull request:
Updated for Boto3.
Added support for listing all application versions without specifying an application name.
Fixed some bugs related to the description variable (if it wasn't used the module could fail because it was expecting it in the response).
Added test playbook.
Test playbook requires the upgraded elasticbeanstalk_app module #3 

👍

rohitkothari commented on library/elasticbeanstalk_version.py:
```diff
@@ -79,10 +84,22 @@ def describe_version(ebs, app_name, version_label):
+        versions = ebs.describe_application_versions(
+            ApplicationName=app_name,
+            VersionLabels=version_label)
+        versions = versions["ApplicationVersions"]
```
Any specific reason to override the same `versions` variable value (from response of describe_application_versions to versions["ApplicationVersions"])? This line would strip off the original API call response and reset a part of that response to same variable if that's what is intended here.

sidewinder12s commented:
@rohitkothari So do we think this is probably good to go now?

```
▶ ansible-playbook elasticbeanstalk_version-tests.yml --extra-vars "app_version_bucket=(HIDDEN)"
 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ***************************************************************

TASK [set_fact (OK)] ***********************************************************
ok: [localhost]

TASK [assert] ******************************************************************
ok: [localhost]

TASK [Create Application for Applicaiton version testing (CHANGED)] ************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "created_application": {
        "Applications": [
            {
                "ApplicationName": "elasticbeanstalk_version-testing",
                "ConfigurationTemplates": [],
                "DateCreated": "2016-11-11T23:10:16.155000+00:00",
                "DateUpdated": "2016-11-11T23:10:16.155000+00:00",
                "Description": "Testing boto3 elasticbeanstalk_version Ansible Module"
            }
        ],
        "changed": true
    }
}

TASK [Create Application Version (CHANGED)] ************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "created_application_version": {
        "ApplicationVersions": [
            {
                "ApplicationName": "elasticbeanstalk_version-testing",
                "DateCreated": "2016-11-11T23:10:17.789000+00:00",
                "DateUpdated": "2016-11-11T23:10:17.789000+00:00",
                "SourceBundle": {
                    "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                    "S3Key": "java-tomcat-v1.zip"
                },
                "Status": "UNPROCESSED",
                "VersionLabel": "Testing"
            }
        ],
        "changed": true
    }
}

TASK [Test Idempotency (OK)] ***************************************************
ok: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [Test Updating Description (CHANGED)] *************************************
changed: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [debug] *******************************************************************
ok: [localhost] => {
    "new_created_application_version": {
        "changed": true,
        "msg": "All items completed",
        "results": [
            {
                "ApplicationVersions": [
                    {
                        "ApplicationName": "elasticbeanstalk_version-testing",
                        "DateCreated": "2016-11-11T23:10:17.789000+00:00",
                        "DateUpdated": "2016-11-11T23:10:19.236000+00:00",
                        "Description": "Test adding a description",
                        "SourceBundle": {
                            "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                            "S3Key": "java-tomcat-v1.zip"
                        },
                        "Status": "UNPROCESSED",
                        "VersionLabel": "Testing"
                    }
                ],
                "_ansible_item_result": true,
                "_ansible_no_log": false,
                "_ansible_parsed": true,
                "changed": true,
                "invocation": {
                    "module_args": {
                        "app_name": "elasticbeanstalk_version-testing",
                        "aws_access_key": null,
                        "aws_secret_key": null,
                        "delete_source": false,
                        "description": "Test adding a description",
                        "ec2_url": null,
                        "profile": null,
                        "region": "us-west-2",
                        "s3_bucket": "elasticbeanstalk-us-west-2-246818788759",
                        "s3_key": "java-tomcat-v1.zip",
                        "security_token": null,
                        "state": "present",
                        "validate_certs": true,
                        "version_label": "Testing"
                    },
                    "module_name": "elasticbeanstalk_version"
                },
                "item": {
                    "ApplicationName": "elasticbeanstalk_version-testing",
                    "ConfigurationTemplates": [],
                    "DateCreated": "2016-11-11T23:10:16.155000+00:00",
                    "DateUpdated": "2016-11-11T23:10:16.155000+00:00",
                    "Description": "Testing boto3 elasticbeanstalk_version Ansible Module"
                }
            }
        ]
    }
}

TASK [Test creating another Application Version (CHANGED)] *********************
changed: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [Test Listing our created Application Version (OK)] ***********************
ok: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [debug] *******************************************************************
ok: [localhost] => {
    "list_application_version": {
        "changed": false,
        "msg": "All items completed",
        "results": [
            {
                "ApplicationVersions": [
                    {
                        "ApplicationName": "elasticbeanstalk_version-testing",
                        "DateCreated": "2016-11-11T23:10:20.325000+00:00",
                        "DateUpdated": "2016-11-11T23:10:20.325000+00:00",
                        "SourceBundle": {
                            "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                            "S3Key": "java-tomcat-v1.zip"
                        },
                        "Status": "UNPROCESSED",
                        "VersionLabel": "Testing-2"
                    },
                    {
                        "ApplicationName": "elasticbeanstalk_version-testing",
                        "DateCreated": "2016-11-11T23:10:17.789000+00:00",
                        "DateUpdated": "2016-11-11T23:10:19.236000+00:00",
                        "Description": "Test adding a description",
                        "SourceBundle": {
                            "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                            "S3Key": "java-tomcat-v1.zip"
                        },
                        "Status": "UNPROCESSED",
                        "VersionLabel": "Testing"
                    }
                ],
                "_ansible_item_result": true,
                "_ansible_no_log": false,
                "_ansible_parsed": true,
                "changed": false,
                "invocation": {
                    "module_args": {
                        "app_name": "elasticbeanstalk_version-testing",
                        "aws_access_key": null,
                        "aws_secret_key": null,
                        "delete_source": false,
                        "description": null,
                        "ec2_url": null,
                        "profile": null,
                        "region": "us-west-2",
                        "s3_bucket": null,
                        "s3_key": null,
                        "security_token": null,
                        "state": "list",
                        "validate_certs": true,
                        "version_label": null
                    },
                    "module_name": "elasticbeanstalk_version"
                },
                "item": {
                    "ApplicationName": "elasticbeanstalk_version-testing",
                    "ConfigurationTemplates": [],
                    "DateCreated": "2016-11-11T23:10:16.155000+00:00",
                    "DateUpdated": "2016-11-11T23:10:16.155000+00:00",
                    "Description": "Testing boto3 elasticbeanstalk_version Ansible Module"
                }
            }
        ]
    }
}

TASK [Test Listing all Application Version (OK)] *******************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "all_app_versions": {
        "ApplicationVersions": [
            {
                "ApplicationName": "elasticbeanstalk_version-testing",
                "DateCreated": "2016-11-11T23:10:20.325000+00:00",
                "DateUpdated": "2016-11-11T23:10:20.325000+00:00",
                "SourceBundle": {
                    "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                    "S3Key": "java-tomcat-v1.zip"
                },
                "Status": "UNPROCESSED",
                "VersionLabel": "Testing-2"
            },
            {
                "ApplicationName": "elasticbeanstalk_version-testing",
                "DateCreated": "2016-11-11T23:10:17.789000+00:00",
                "DateUpdated": "2016-11-11T23:10:19.236000+00:00",
                "Description": "Test adding a description",
                "SourceBundle": {
                    "S3Bucket": "elasticbeanstalk-us-west-2-246818788759",
                    "S3Key": "java-tomcat-v1.zip"
                },
                "Status": "UNPROCESSED",
                "VersionLabel": "Testing"
            }
        ],
        "changed": false
    }
}

TASK [Test listing a non-existent Application (OK)] ****************************
ok: [localhost]

TASK [Display Non-existent Application Application Versions (Should be empty)] *
ok: [localhost] => {
    "Non_existent": {
        "changed": false,
        "output": "Version not found"
    }
}

TASK [Delete the Application Version (CHANGED)] ********************************
changed: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [Idempotent delete the Application Version (OK)] **************************
ok: [localhost] => (item={u'ApplicationName': u'elasticbeanstalk_version-testing', u'ConfigurationTemplates': [], u'DateUpdated': u'2016-11-11T23:10:16.155000+00:00', u'Description': u'Testing boto3 elasticbeanstalk_version Ansible Module', u'DateCreated': u'2016-11-11T23:10:16.155000+00:00'})

TASK [Delete the application (CHANGED)] ****************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=19   changed=6    unreachable=0    failed=0

```

sidewinder12s commented:
👍  Still good I think based on the above output, last push just updated the test playbook a bit to display un-itemized elasticbeanstalk_versions output.

rohitkothari commented:
:+1:
